### PR TITLE
rust: Support http/1.1

### LINKED ISSUE #14: ch1: Support HTTP/1.1
Along with Host, send the Connection header in the request function with the value close. Your browser can now declare that it is using HTTP/1.1. Also add a User-Agent header. Its value can be whatever you want—it identifies your browser to the host. Make it easy to add further headers in the future

- [x] Python
- [ ] Rust

### PR DESCRIPTION
**Pull request by corona10**
closes: #14

### CODE REVIEW

File: rust/src/lib.rs [hunk 1]
```diff
@@ -138,12 +167,16 @@ pub mod http {
...
         write!(
             stream,
-            "GET {} HTTP/1.0\r
+            "GET {} HTTP/1.1\r
 Host: {}\r
+Connction: close\r
+User-Agent: Mozilla/5.0 ({})\r
 Accept-Encoding: gzip,deflate\r
 \r
 ",
```
Comment (youknowone) [inline rust/src/lib.rs]: This is not a new decision from this PR, but you may want to use multiline whitespace escape machanism
https://doc.rust-lang.org/rust-by-example/std/str.html#literals-and-escapes

File: rust/src/lib.rs [hunk 2]
```diff
@@ -180,34 +212,42 @@ Accept-Encoding: gzip,deflate\r
             let (header, value) = split2(&line, ":").ok_or(MALFORMED_RESPONSE)?;
             let header = header.to_ascii_lowercase();
             let value = value.trim();
-            if header == "content-encoding" {
-                encoding = value.parse().map_err(|_| UNSUPPORTED_ENCODING)?;
-            }
             headers.insert(header, value.to_string());
         }
 
-        // 11. Read body
```
Comment (youknowone) [inline rust/src/lib.rs:189]: I feel like this `11. Read body` comment is intended to be alive somewhere for educational purpose but it is now deleted.

File: rust/src/lib.rs [hunk 3]
```diff
@@ -90,6 +91,34 @@ pub mod http {
         Some((split.next()?, split.next()?))
     }
 
+    fn decompress<R: Read>(reader: &mut BufReader<R>, encoding: ContentEncoding) -> Vec<u8> {
```
Comment (youknowone) [inline rust/src/lib.rs]: if we don't care about performance but care about the readability of code, having a 'decompressed reader' might be simple and more comprehensible.

```rust
fn decompressor<R: Read>(reader: R, encoding: ContentEncoding) -> Box<dyn Read> {
    match encoding {
        ContentEncoding::Gzip => Box::new(GzDecoder::new(reader)),
        ContentEncoding::Deflate => Box::new(DeflateDecoder::new(reader)),
        ContentEncoding::Identity => Box::new(reader),
   }
}
```
(edited)

Then, `decompress(reader, encoding).read_to_end(&mut body)` will work.

File: rust/src/lib.rs [hunk 4]
```diff
@@ -180,34 +212,42 @@ Accept-Encoding: gzip,deflate\r
...
+        let body = match headers.get("transfer-encoding") {
+            Some(encoding) => {
+                let mut unchunked = Vec::new();
+                if "chunked".eq_ignore_ascii_case(encoding) {
+                    loop {
+                        let mut line = String::new();
+                        reader
+                            .read_line(&mut line)
+                            .map_err(|_| MALFORMED_RESPONSE)?;
+                        let n_bytes = i64::from_str_radix(line.trim_end(), 16).unwrap_or(0);
```
Comment (youknowone) [inline rust/src/lib.rs]: i64 or u64?

File: rust/src/lib.rs [hunk 5]
```diff
@@ -180,34 +212,42 @@ Accept-Encoding: gzip,deflate\r
...
+                        if n_bytes == 0 {
+                            break;
+                        }
+                        let mut chunk = vec![0u8; n_bytes as usize];
+                        reader
+                            .read_exact(&mut chunk)
+                            .map_err(|_| MALFORMED_RESPONSE)?;
+                        reader.read_exact(&mut vec![0u8; 2]).unwrap();
+                        unchunked.write_all(&chunk).unwrap();
+                    }
```
Comment (youknowone) [inline rust/src/lib.rs]: if you want, writing `ChunkReader` with `impl Read for ChunkReader` will provide a nice way to make `Read` to `Read` conversion.

Comment (corona10) [reply rust/src/lib.rs]: I will try

**Review by youknowone — APPROVED** [review]

**Pull request merged**